languages: add svelte support

happysalada opened the pull request:
I'm opening this prospecive PR just to see if I haven't made anything stupid.

I've added the tree-sitter gitmodule
added all the queries file that were in the tree-sitter repository.

I need to update my nix to include the latest PR that downloads the submodules automatically.
I should have time to test this in the next few days.

kirawi commented on runtime/queries/svelte/folds.scm:
```diff
@@ -0,0 +1,11 @@
+[
```
Folds are not supported.

kirawi commented on runtime/queries/svelte/highlights.scm:
```diff
@@ -0,0 +1,68 @@
+; Special identifiers
```
Some of these need to be converted into Helix's scopes: https://docs.helix-editor.com/themes.html

kirawi replied:
@archseer Scopes look fine to me now; do we want to also add the `none` scope?

kirawi commented on runtime/queries/svelte/indents.scm:
```diff
@@ -0,0 +1,18 @@
+[
```
Indents should be defined using an `indents.toml` file instead.

kirawi reviewed (changes requested):
(By the way, you're missing the final newline on a few of these files).

happysalada commented:
I've added the newlines, and switched from scm to toml for the indent file.
Regarding the highlights changes, understood. I need to test this to make meaningful changes. I'll update as soon as I could try it.

Thank you for the review!

archseer commented on runtime/queries/svelte/injections.scm:
```diff
@@ -0,0 +1,26 @@
+; injections.scm
+; --------------
```
Injections should follow https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection

archseer commented on runtime/queries/svelte/indents.toml:
```diff
@@ -0,0 +1,18 @@
+  "await_statement"
+]
+
+branch = [
```
There is no branch or ignore keys, you might want to add the closing tag as outdent instead

happysalada replied:
changed

archseer commented on runtime/queries/svelte/highlights.scm:
```diff
@@ -0,0 +1,68 @@
+;--------------------
+
+; TODO:
+((element (start_tag (tag_name) @_tag) (text) @text.title)
```
Note: these text tags should likely be https://www.sublimetext.com/docs/scope_naming.html#markup but can be changed in a follow-up since there's a few other grammars to fix too.

happysalada replied:
I've changed to the related markup tags

archseer commented on languages.toml:
```diff
@@ -277,6 +277,15 @@ roots = []
+file-types = ["svelte"]
+roots = []
+indent = { tab-width = 2, unit = "  " }
+language-server = { command = "svelteserver" }
```
Did you test this? It probably needs `args = ["--stdio"]` based on what I've seen elsewhere

happysalada replied:
Not yet, I wanted to make a prospective PR to see if I had misunderstood things.
it seems you are right here
https://github.com/sveltejs/language-tools/blob/master/packages/language-server/src/server.ts#L67

thank you for the review!

kirawi commented on runtime/queries/svelte/highlights.scm:
```diff
@@ -0,0 +1,68 @@
+[
+  "{"
+  "}"
+] @punctuation.separator
```
Should be `punctuation.bracket`

kirawi commented on runtime/queries/svelte/highlights.scm:
```diff
@@ -0,0 +1,68 @@
+  ":"
+  "/"
+  "@"
+] @punctionation.definition.tag
```
Typo `punctionation` vs. `punctuation`. I also think the brackets should be `punctuation.bracket`. Not sure about the rest.

happysalada commented:
Just to give you more context, svelte is a bit like vue, it's a file that contains a mix of html and javascript (or typescript).
Here is how it looks like in a hypothetical inefficient editor
(Note that there are some specificities around `{` brackets in the html)

<img width="1280" alt="Screen Shot 2021-09-10 at 11 13 46" src="https://user-images.githubusercontent.com/5317234/132788290-8ea15033-119c-4482-8165-863549d59ea5.png">
<img width="1280" alt="Screen Shot 2021-09-10 at 11 14 03" src="https://user-images.githubusercontent.com/5317234/132788326-9f2fd495-d3dc-4d41-8e27-6ebd76558b70.png">

happysalada commented:
Ok, I just had time to test this, it doesn't work at all.
Svelte is mostly injections, so I'll have a look at fixing that file first.
I'm not sure why the lsp doesn't do anything though. Not sure how to activate/debug.

archseer commented:
LSP: Is the binary in your $PATH? Run helix with the `-v` flag and post the log file.

happysalada commented:
binary is in my path
```
❯ which svelteserver
/Users/raphael/.nix-profile/bin/svelteserver
```

it seems to run actually, it just doesn't display anything.
Here are the logs
[helix.log](https://github.com/helix-editor/helix/files/7159128/helix.log)

archseer commented:
> binary is in my path
> 
> ```
> ❯ which svelteserver
> /Users/raphael/.nix-profile/bin/svelteserver
> ```
> 
> it seems to run actually, it just doesn't display anything.
> Here are the logs
> [helix.log](https://github.com/helix-editor/helix/files/7159128/helix.log)

It seems to load just fine, are you able to get any completions using ctrl-x in insert mode? Or `k` over some item to see some documentation etc. I can see the LSP initialization but no interaction with the server.

happysalada commented:
I wasn't aware of `ctrl-x`.
it does work indeed. `ctrl-x` generates completion. typing some wrong code also gives errors in the editor.
There are just the injections left then. Let me update this in the next few days.

happysalada commented:
I've modified the injections to be of the correct format (I think), however those are not working. a svelte file is not being syntax hyghlighted.
I've checked the logs with `-v` and there are no error or indication that something is going wrong.
Not really sure what to try next.

archseer commented:
Can you post a sample `.svelte` file? I'll give it a go.

happysalada commented:
[index.svelte.txt](https://github.com/helix-editor/helix/files/7166121/index.svelte.txt)

(the normal extension is just .svelte , I added a .txt to enable github upload, feel free to remove the .txt once you've downloaded it)
That should be a fairly standard one. Let me know if I can help in any way.

archseer commented:
Fixed in https://github.com/helix-editor/helix/commit/dd0b15e1f1540d6ca8c58594be302c66005d755c! It worked on `css` but not for typescript or javascript. I'll let you do a quick test, but other than that I think we're ready for merge.

happysalada commented:
The injections didn't work on a style of script attribute for me.
I'm happy to merge though, as having a language server is an infinite improvement.
Plus now that I know where the injections are happening in the code, it gives me a strong incentive to go have a look and figure it out.

Thanks a lot for the help and the time spent on this!

archseer commented:
> The injections didn't work on a style of script attribute for me.

They don't work for you? `<script lang="ts">` was highlighted correctly for me after the patch in master.

happysalada commented:
I rebased this branch on master this morning, made sure I had my branch running locally.
I can confirm that I don't get highlights inside script or style tags.
I've turned verbosity on and saw nothing in the logs either.
I'm planning on having a look at the commit you pointed out and how the highlighting works in the next few weeks.

archseer commented:
Make sure you `cargo clean -p helix-syntax` and clear out `runtime/grammars/` before building, maybe that will fix it.

happysalada commented:
cool, let me check.

happysalada commented:
Oh no, just realized that can't be it.
I'm building with the flake directly. So everytime, it compiles from scratch (some of the dependencies are cached of course).

archseer commented:
![pos](https://user-images.githubusercontent.com/1372918/133707612-b347dfbb-0e27-4577-9fe8-37b4b059e5b0.png)

Here's the file you provided for example, after building on this branch from scratch.

happysalada commented:
this looks beautiful!

ok let me try to build without the flake then.

happysalada commented:
Thanks a lot, local compilation did work, it must be a problem with the flake (or more probable, my usage of it).
